pelican/log.py

```diff
@@ -1,142 +1,446 @@
 import logging
 from collections import defaultdict
+from contextlib import contextmanager
 
 from rich.console import Console
 from rich.logging import RichHandler
 
-__all__ = [
-    'init'
-]
 
 console = Console()
 
+# handles messages logged from anywhere within the pelican namespace.
+PKG_LOGGER = logging.getLogger('pelican')
+
+
+def severity_from_name(string_val):
+    """Convert string value to appropriate severity (AKA log level)."""
+    if not string_val:
+        return logging.NOTSET
+    level_name = string_val.upper().rstrip('S')
+    try:
+        return logging.getLevelNamesMapping()[level_name]
+    except AttributeError:  # Python < 3.11
+        return logging._nameToLevel.copy()[level_name]
+
+
+@contextmanager
+def temp_filter(logger, flt):
+    """
+    Context in which a custom filter is temporarily applied to logger.
+    """
+    try:
+        logger.addFilter(flt)
+        yield flt
```

Comment:
The logger doesn't need to be yielded: the caller has already provided it, and it isn't even always referenced within the context.

The filter, however, is intended to be temporary; yielding the filter allows the caller to generate the filter on entering the context and have it garbage collected on exiting the context.  It essentially allows this:

```python
from pelican import log
import logging


with log.temp_filter(logging.getLogger("test_module", MyCustomFilter()) as tmp_filter:
   test_module.do_stuff()
 

# outer scope needs not keep reference to filter; it can be garbage collected
```

This is an interface; while not used called within this module, it is actually used in several places in testing, and can be useful in scenarios where a filter is needed only for a subsection of code.

Reply:
What you yield has nothing to do with references. You don't need a reference to either of them outside `with` block regardless of what is yielded. Context manager is already doing that during `with`. You can `yield None` or bare `yield` if you wanted. `yield` only defines the value that will be assigned to the name after `as`. 

What I'm saying is, you're far more likely to do something with the logger inside the `with` block than a filter or a handler. i.e. your example could become:

```python
with log.temp_filter(logging.getLogger("test_module"), MyCustomFilter()) as logger:
    logger.warning(...)  # 
    test_module.do_stuff()
```